sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatch.java

```diff
@@ -0,0 +1,322 @@
+  private static final int DEFAULT_CAPACITY = 1 << 16;
+  private static final long DEFAULT_PAGE_SIZE = 64 * 1024 * 1024;
+
+  private final StructType keySchema;
+  private final StructType valueSchema;
+  private final int capacity;
+  private int numRows = 0;
+
+  // Staging row returned from getRow.
+  final UnsafeRow keyRow;
+  final UnsafeRow valueRow;
+
+  // ids for current key row and value row being retrieved
+  private int keyRowId = -1;
+
+  // full addresses for key rows and value rows
+  private long[] keyOffsets;
+
+  // if all data types in the schema are fixed length
+  private boolean allFixedLength;
+  private int klen;
+  private int vlen;
+  private int recordLength;
+
+  private MemoryBlock currentAndOnlyPage = null;
+  private Object currentAndOnlyBase = null;
+  private long recordStartOffset;
+  private long pageCursor = 0;
+
+  public static RowBasedKeyValueBatch allocate(StructType keySchema, StructType valueSchema,
+                                               TaskMemoryManager manager) {
+    return new RowBasedKeyValueBatch(keySchema, valueSchema, DEFAULT_CAPACITY, manager);
+  }
+
+  public static RowBasedKeyValueBatch allocate(StructType keySchema, StructType valueSchema,
+                                               TaskMemoryManager manager, int maxRows) {
+    return new RowBasedKeyValueBatch(keySchema, valueSchema, maxRows, manager);
+  }
+
+  public int numRows() { return numRows; }
+
+  public void close() {
+    if (currentAndOnlyPage != null) {
+      freePage(currentAndOnlyPage);
+      currentAndOnlyPage = null;
+    }
+  }
+
+  private boolean acquireNewPage(long required) {
+    try {
+      currentAndOnlyPage = allocatePage(required);
+    } catch (OutOfMemoryError e) {
+      return false;
+    }
+    currentAndOnlyBase = currentAndOnlyPage.getBaseObject();
+    Platform.putInt(currentAndOnlyBase, currentAndOnlyPage.getBaseOffset(), 0);
+    pageCursor = 4;
+    recordStartOffset = pageCursor + currentAndOnlyPage.getBaseOffset();
+
+    return true;
+  }
+
+  private long getKeyOffsetForFixedLengthRecords(int rowId) {
+    return recordStartOffset + rowId * recordLength + 8;
+  }
+
+  public UnsafeRow appendRow(Object kbase, long koff, int klen,
+                             Object vbase, long voff, int vlen) {
+    final long recordLength = 8 + klen + vlen + 8;
+    // if run out of max supported rows or page size, return null
+    if (numRows >= capacity || currentAndOnlyPage == null
+            || currentAndOnlyPage.size() - pageCursor < recordLength) {
+      return null;
+    }
+
+    final Object base = currentAndOnlyBase;
+    long offset = currentAndOnlyPage.getBaseOffset() + pageCursor;
+    final long recordOffset = offset;
+    if (!allFixedLength) { // we only put lengths info for variable length
+      Platform.putInt(base, offset, klen + vlen + 4);
+      Platform.putInt(base, offset + 4, klen);
+    }
+    offset += 8;
```

Comment:
We seem to be wasting space here for fixed length. As discussed offline, let's just have the fixed length case in a different class that extends this class (or a common base class). This would significantly make it easier to reason about the 2 cases. As a side affect, we can then also set klen, vlen etc. to final for the fixed length case.